# ***Titanic - Machine Learning from Disaster***

***Ćwiczenie to polegało na zrealizowaniu zadania konkursowego ze strony Kaggle dotyczącego określenia ile i którzy pasażerowie przetrwają katastrofę Titanica. Dane do zadania są zawarte w dwóch plikach, train.csv i test.csv. Pliki te zawierają tabelę z informacje o pasażerach, takie jak np. wiek, płeć, rodzaj biletu itp. Model do nauki ma wykorzystać dane z pliku train.csv a następnie ma przeanalizować plik test.csv i określić kto może przetrwać katastrofę Titanica.***

In [85]:
# This Python 3 environment comes with many helpful analytics libraries installed
# It is defined by the kaggle/python Docker image: https://github.com/kaggle/docker-python
# For example, here's several helpful packages to load

import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)

# Input data files are available in the read-only "../input/" directory
# For example, running this (by clicking run or pressing Shift+Enter) will list all files under the input directory

import os
for dirname, _, filenames in os.walk('/kaggle/input'):
    for filename in filenames:
        print(os.path.join(dirname, filename))

# You can write up to 20GB to the current directory (/kaggle/working/) that gets preserved as output when you create a version using "Save & Run All" 
# You can also write temporary files to /kaggle/temp/, but they won't be saved outside of the current session

/kaggle/input/titanic/train.csv
/kaggle/input/titanic/test.csv
/kaggle/input/titanic/gender_submission.csv


In [86]:
train_data = pd.read_csv("/kaggle/input/titanic/train.csv")
train_data.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

***Wynikiem działa powyższego kodu jest wypisanie z tabeli zawierającej dane do nauki modelu pięciu pierwszych elementami tabeli.***

In [87]:
test_data = pd.read_csv("/kaggle/input/titanic/test.csv")
test_data.head()

PassengerId  Pclass                                          Name     Sex  \
0          892       3                              Kelly, Mr. James    male   
1          893       3              Wilkes, Mrs. James (Ellen Needs)  female   
2          894       2                     Myles, Mr. Thomas Francis    male   
3          895       3                              Wirz, Mr. Albert    male   
4          896       3  Hirvonen, Mrs. Alexander (Helga E Lindqvist)  female   

    Age  SibSp  Parch   Ticket     Fare Cabin Embarked  
0  34.5      0      0   330911   7.8292   NaN        Q  
1  47.0      1      0   363272   7.0000   NaN        S  
2  62.0      0      0   240276   9.6875   NaN        Q  
3  27.0      0      0   315154   8.6625   NaN        S  
4  22.0      1      1  3101298  12.2875   NaN        S

***Wynikiem działa powyższego kodu jest wypisanie z tabeli zawierającej dane na których będziemy przeprowadzać analizę pięciu pierwszych elementami tabeli.***

In [88]:
women = train_data.loc[train_data.Sex == 'female']["Survived"]
rate_women = sum(women)/len(women)
men = train_data.loc[train_data.Sex == 'male']["Survived"]
rate_men = sum(men)/len(men)

print("% of women who survived:", rate_women)
print("% of men who survived:", rate_men)

% of women who survived: 0.7420382165605095
% of men who survived: 0.18890814558058924
314
233
577
109


***Powyższy kod podzielił wszystkich pasażerów ze względu na ich płeć i wygenerował nam procentową ilość kobiet i mężczyzn którzy przeżyli. 
Dane pasażerów były brane z pliku train.csv czyli z tego przeznaczonego do nauki modelu.
Możemy zauważyć że katastrofę przeżyło znacznie więcej kobiet bo około 74% natomiast mężczyzn przeżyło niecałe 19%.***

In [89]:
from sklearn.ensemble import RandomForestClassifier

y = train_data["Survived"]
features = ["Pclass", "Sex", "SibSp", "Parch"]

***Powyższy kod pozwala nam zaimportować plik z modelem nauki maszynowej. 
Dodatkowo określam w tym miejscy również argumenty jakie będą brane pod uwagę podczas nauki.***

In [90]:
X = pd.get_dummies(train_data[features])
X_test = pd.get_dummies(test_data[features])

***Do zmiennej "X" wprowadzam dane z pliku "train.csv" czyli dane przeznaczone do nauki modelu.
Natomiast do zmiennej "X_test" wprowadzam dane z pliku "test.csv" czyli dane które będą poddawane analizie.***

In [91]:
model = RandomForestClassifier(n_estimators=25, max_depth=5, random_state=1)
model.fit(X, y)
predictions = model.predict(X_test)

output = pd.DataFrame({'PassengerId': test_data.PassengerId, 'Survived': predictions})
output.to_csv('my_submission.csv', index=False)
print("Your submission was successfully saved!")

Your submission was successfully saved!


***W kodzie powyżej definiuję model oraz wykonuję prognozę wyników.***

In [92]:
score = model.score(X,y)
print(score)

my_data = pd.read_csv("./my_submission.csv")
my_data.head()

0.8159371492704826


PassengerId  Survived
0          892         0
1          893         1
2          894         0
3          895         0
4          896         1

***Wykorzystując tą metodę oraz dobrane parametry uzyskałem wynik na poziomi 0.815937***